### DISCUSSION TITLE
$watch array output the same value between old and new.

### DISCUSSION THREAD

**Discussion by dinyo** [alpinejs/alpine#3293] [7. Bugs (archived)]
maybe i am doing it wrong.
doing $watch on an array and edit the input value in loop, (console) output the same value between old and new.

code:
https://codepen.io/dinyo/pen/gOvJowE

thank you.

edit: turn out deep watch give wrong oldvalue. even example from documentation are wrong. tested from 3.8 to 3.10

**Comment by joeszetocodeman** [comment]
I think this is not an issue of alpine, even use that pure javascript will get the same result.
you setting the "field" in an object, and the object was pushed in an array, when you update the "field" value in this object
no-mater the new val or the oldval of the array is pointing to the same object in memory heap.

**Comment by dinyo** [comment]
hmmm is that so? anyway the alpine example give the wrong output (not the same as the doc claim).

doc:
https://alpinejs.dev/magics/watch#deep-watching

codepen:
https://codepen.io/dinyo/pen/eYVaKgj?editors=1111

thanks for reply tho.

**Comment by dinyo** [comment]
i think the problem in line 20 and line 23 in $watch.js
https://github.com/alpinejs/alpine/blob/763e287842da1d9e0a32c52015cfd5c8f33dbac7/packages/alpinejs/src/magics/$watch.js

`oldValue = value`
because of deep watch, the value is a proxy object, then the oldvalue refer to the proxy object.

currently i am using this on my local build.
`oldValue = JSON.parse(JSON.stringify(value))`

**Comment by puckowski** [comment]
Issue still seems to be present in latest main (08/23/2022 5:52 EST). See CodePen:
https://codepen.io/puckowski/pen/WNzWRQw?editors=1111

I think https://github.com/alpinejs/alpine/pull/3046 addresses the issue with additional Cypress tests added.

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/alpinejs/alpine/discussions/1423

**Comment by puckowski** [comment]
Issue seems present in v3.12.2:
https://codepen.io/puckowski/pen/gOQpYOB?editors=1111

Issue is resolved in:
https://codepen.io/puckowski/pen/poQJzzv?editors=1111

See pull here:
https://github.com/alpinejs/alpine/pull/3617

Small addition to bundle size, should cover most cases, with additional watch tests to confirm the solution works.

**Comment by ekwoka** [comment]
I'm always still surprised people use $watch.

Also, surprised people would actually use the old value for object in any meaningful way....
